### ISSUE TITLE
no-wrap-func false positive on callback

### ISSUE THREAD

**Issue by zloirock** [eslint/eslint#2477]
```js
// on
var foo = new (A || B)(function(){});
// or
var bar = new (getConstructor(this))(function(){});
```
eslint 0.21.0

**Comment by ilyavolodin** [comment]
Can you provide console output for those errors please?

**Comment by zloirock** [comment]
For example,
```js
  161:59  error  Wrapping non-IIFE function literals in parens is unnecessary  no-wrap-func
```
for [this](https://github.com/zloirock/core-js/blob/master/modules/es6.promise.js#L161) line.

**Comment by nzakas** [comment]
Looks like a parenthesized callee confuses the rule.

**Comment by jrvidal** [comment]
Working on this

**Comment by jrvidal** [comment]
I started to work on a solution but I realized that I was wildly overlapping with `no-extra-parens`. I'm a little bit confused about what the scope of both rules should be. Isn't `no-wrap-func` just a particular case of the former?

**Comment by michaelficarra** [comment]
Yes. For people that want unnecessary parentheses everywhere except around function expressions.

**Comment by jrvidal** [comment]
Here are some examples:

```js
// (a) Throws `no-extra-parens`
((function f() {})).length;

// (b) It doesn't throw anything
((function () {}))(1);

// (c) Throws `no-wrap-func` and `no-extra-parens`.
var x = (function g() {});
```

If `no-wrap-func` is supposed to be a particular case of `no-extra-parens`, shouldn't both rules throw for all the cases above?

If that is the case, it may make more sense to merge `no-wrap-func` into `no-extra-parens` as an option (`"all"`, `"functions-only"`).

**Comment by michaelficarra** [comment]
`no-extra-parens` does not warn on IIFEs so that it does not overlap with *this rule's original intent*. Since then, however, this rule has changed.

**Comment by nzakas** [comment]
We try to avoid overlaps in warnings, sometimes it happens without us noticing.

It might make sense to combine at this point, but I'm not sure of the correct option names.

**Comment by zloirock** [comment]
Still not fixed in 0.22 =\

**Comment by jrvidal** [comment]
I think the intention is to deprecate `no-wrap-func` and use the "functions" option in [`no-extra-parens`](http://eslint.org/docs/rules/no-extra-parens) instead

**Comment by zloirock** [comment]
@jrvidal I don't use `no-wrap-func`, `no-extra-parens` enough for me, but after removing `no-wrap-func` from `.eslintrc` I still have this error.

**Comment by jrvidal** [comment]
@zloirock Could you provide a test case? I am unable to reproduce it with your examples above.

**Comment by zloirock** [comment]
@jrvidal https://github.com/zloirock/core-js :) after removing `"no-wrap-func": 0` from `.eslintrc` and updating to `0.22` - still this error.

**Comment by jrvidal** [comment]
@zloirock Oh, sorry, I didn't understand you. Yes, `no-wrap-func` hasn't been fixed. If you have `no-extra-parens` enabled, you'll get warnings for unnecessary function wrapping. Meanwhile, `no-wrap-func` is still enabled by default. I'll open an issue to deprecate it properly.

**Comment by zloirock** [comment]
@jrvidal looks like I didn't understand you too. I thought, you wrote `no-wrap-func` was deprecated in previous commits :)
